### DISCUSSION TITLE
Potential Modding Assistance: File Formats

### DISCUSSION THREAD

**Discussion by AugustRenders** [DarkflameUniverse/DarkflameServer#1137] [General]
This discussion's purpose is to document all of the file formats used in Lego Universe, as well as their purposes and how they can be accessed. I am digging into the Lego Universe game files, and will continue updating this page as I go.

### List of File Formats:

.cfg
Purpose: To store configurations. (We have to replace it with an edited version when linking our LU software to a Darkflame server.)
Software: Unknown

.dds:
Purpose: These are image files used for texture maps and the User Interface.
Software: They can be opened and saved with GIMP.

.xml
Purpose: Examination of the code and file placement tells me that it's for User Interface functionality.
Software: Unknown, was able to open in Notepad.

.gfx
Purpose: Containing graphical assets.
Software: Unknown

.fsb:
Purpose: These are audio files used to store sound effects and music.
Software: While there are some programs that can open .fsb files, FMOD Studio is the only one that can export them, as it is a proprietary file format.

.fev
Purpose: Contains events, assets, parameters and other data that defines the interactive experience of audio within a game.
Software: FMOD Studio only

.lua:
Purpose: These are scripts that use the Lua programming language.
Software: Possibly Notepad++

.scm
Purpose: Commonly used as a file format for scripting in games.
Software: Unknown

.bik
Purpose: These are video files used for things like cutscenes. (i.e. The Venture Explorer falling into the Maelstrom vortex.)
Software: Can be opened with VLC media player. This type of file format can only be exported by Bink Video SDK.

.txt
Purpose: To contain text. (i.e. Minifigure Names)
Software: Can be opened in Notepad.

.nif
Purpose: To contain models.
Software: NifSkope. Blender plugins are available.

.kfm
Purpose: To contain model animations. (Rig data included.)
Software: There are Blender plugins for importing and exporting these files.

.kf
Purpose: Same as above.
Software: Whether Blender plugins are available is unknown.

.lxfml
Purpose: Lego Digital Designer files.
Software: Supported by Lego Digital Designer, as well as BrickLink Studio and LDCad.

.hkx
Purpose: To contain physics simulations created by the Havok physics engine.
Software: Blender plugins are available to open these files.

.lvl
Purpose: To contain game level files.
Software: Gamebryo (assumed)

.settings
Purpose: Likely used for game configuration settings adjusted by the player.
Software: Unknown.

.lutriggers
Purpose: Short for "Lego Universe Triggers." Functionality unknown.
Software: None. Proprietary format used.

.luz
Purpose: Unknown
Software: Unknown

.evc
Purpose: Unknown
Software: Unknown

.raw
Purpose: Unknown
Software: Unknown

.ast
Purpose: Unknown
Software: Unknown

.aud
Purpose: Unknown, possibly audio.
Software: Unknown

.nff
Purpose: Unknown. It's stored under "fonts," but research suggests a different purpose.
Software: Unknown

.g
Purpose: Stored under "brickprimitives > lod0, lod2, lod3." (L.O.D. is short for "Level of Detail," in relation to models.)
Software: Unknown.

.h
Purpose: Unconfirmed, but this format is often associated with "header files."
Software: Unknown

.etk
Purpose: Unknown
Software: Unknown

.dll
Purpose: Unknown, described as an "application extension" under "Type" in the Windows File Explorer.
Software: Unknown

.default
Purpose: Unknown
Software: Unknown

.state
Purpose: Unknown
Software: Unknown

.psb
Purpose: Unknown. Stored under res > forkp > effects. Then go into the folders that follow.
Software: Unknown. Research suggests that it could be viewed in Adobe Photoshop.

**Comment by EmosewaMC** [comment]
All this information can be found in the [docs](https://docs.lu-dev.net/en/latest/index.html) linked in the [README](https://github.com/DarkflameUniverse/DarkflameServer#development-documentation).
